ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/CommandDeduplication.scala

```diff
@@ -3,90 +3,205 @@
-import com.daml.ledger.participant.state.kvutils.store.{DamlCommandDedupValue, DamlStateValue}
+import com.daml.ledger.participant.state.kvutils.store.DamlCommandDedupValue.TimeCase
 import com.daml.ledger.participant.state.kvutils.store.events.{
+  DamlSubmitterInfo,
   DamlTransactionRejectionEntry,
   Duplicate,
 }
+import com.daml.ledger.participant.state.kvutils.store.{
+  DamlCommandDedupValue,
+  DamlLogEntry,
+  DamlStateValue,
+  PreExecutionDeduplicationBounds,
+}
+import com.daml.ledger.participant.state.kvutils.{Conversions, Err}
 import com.daml.logging.LoggingContext
 
 private[transaction] object CommandDeduplication {
 
-  def overwriteDeduplicationPeriodWithMaxDurationStep(
-      defaultConfig: Configuration
-  ): Step = new Step {
-    override def apply(context: CommitContext, input: DamlTransactionEntrySummary)(implicit
-        loggingContext: LoggingContext
-    ): StepResult[DamlTransactionEntrySummary] = {
-      val (_, currentConfig) = getCurrentConfiguration(defaultConfig, context)
-      val submission = input.submission.toBuilder
-      submission.getSubmitterInfoBuilder.setDeduplicationDuration(
-        buildDuration(currentConfig.maxDeduplicationTime)
-      )
-      StepContinue(input.copyPreservingDecodedTransaction(submission.build()))
-    }
-  }
-
   /** Reject duplicate commands
     */
-  def deduplicateCommandStep(rejections: Rejections): Step = new Step {
-    def apply(
-        commitContext: CommitContext,
-        transactionEntry: DamlTransactionEntrySummary,
-    )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
-      commitContext.recordTime
-        .map { recordTime =>
-          val dedupKey = commandDedupKey(transactionEntry.submitterInfo)
-          val dedupEntry = commitContext.get(dedupKey)
-          if (dedupEntry.forall(isAfterDeduplicationTime(recordTime.toInstant, _))) {
-            StepContinue(transactionEntry)
-          } else {
-            rejections.reject(
-              DamlTransactionRejectionEntry.newBuilder
-                .setSubmitterInfo(transactionEntry.submitterInfo)
-                // No duplicate rejection is a definite answer as the deduplication entry will eventually expire.
-                .setDefiniteAnswer(false)
-                .setDuplicateCommand(Duplicate.newBuilder.setDetails("")),
-              "the command is a duplicate",
-              commitContext.recordTime,
+  def deduplicateCommandStep(rejections: Rejections): Step =
+    new Step {
+      def apply(
+          commitContext: CommitContext,
+          transactionEntry: DamlTransactionEntrySummary,
+      )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
+        val commandDeduplicationDuration =
+          if (transactionEntry.submitterInfo.hasDeduplicationDuration)
+            parseDuration(transactionEntry.submitterInfo.getDeduplicationDuration)
+          else
+            throw Err.InternalError(
```

Comment:
What if nothing is specified?

Reply:
That's not supported as of yet. It must always be specified as a duration

Reply:
I don't think we have plans for supporting anything else than a duration for now. If the ledger client specifies an offset, then it is up to the participant to convert it into a suitable duration.